### PR TITLE
[cryptofuzz] Add OpenSSL 1.0.2 and 1.1.0 targets

### PR DESCRIPTION
**Pull request by guidovranken**
There are a bunch of bugs in 1.0.2 and 1.10 that will cause crashes, like https://github.com/openssl/openssl/issues/8980 and https://github.com/openssl/openssl/issues/8972

I still have to create bug reports for the other ones.

You can either merge it now and let ClusterFuzz detect these bugs, or wait until they are all fixed, whichever you prefer.

CC @mattcaswell @kroeckx 

I will remove these targets when they reach end-of-life later this year.

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
I'll merge this once the travis bugs are fixed.

**Comment by guidovranken** [comment]
Part of the EverCrypt build will not succeed, this is expected, hence we compile with ```make -j$(nproc) || true```: https://github.com/google/oss-fuzz/blob/5cfae06b7e9d2672b7afec2e670ccd186553a5f8/projects/cryptofuzz/build.sh#L108

Apparently Travis trips over this.. @s-zanella can the EverCrypt build script be changed to succeed entirely with sanitizers enabled?

**Comment by s-zanella** [comment]
@guidovranken I pushed a change to [the PR I've opened in your fork](https://github.com/guidovranken/oss-fuzz/pull/1) that addresses this.

You can save a lot of time (and space in logs) by only compiling libcrypto.a in OpenSSL builds. Unfortunately, there is no recipe that works across all versions, but I believe `make build_generated libcrypto.a` works for OpenSSL 1.1.x and `make build_crypto` works for OpenSSL 1.0.2 (see https://github.com/openssl/openssl/issues/4597).

File: projects/cryptofuzz/build.sh [hunk 1]
```diff
@@ -272,3 +272,117 @@ cp $SRC/cryptofuzz/cryptofuzz $OUT/cryptofuzz-boringssl-noasm
...
+cd $SRC;
+unzip OpenSSL_1_1_0-stable.zip
+
+if [[ $CFLAGS != *sanitize=memory* ]]
+then
+    # Compile Openssl 1.1.0 (with assembly)
+    cd $SRC/openssl-OpenSSL_1_1_0-stable/
+    ./config --debug enable-md2 enable-rc5 $CFLAGS
+    make depend
+    make -j$(nproc)
```
Comment (s-zanella) [inline projects/cryptofuzz/build.sh]:
```suggestion
    make -j$(nproc) build_generated libcrypto.a
```

File: projects/cryptofuzz/build.sh [hunk 2]
```diff
@@ -272,3 +272,117 @@ cp $SRC/cryptofuzz/cryptofuzz $OUT/cryptofuzz-boringssl-noasm
...
+    cp $SRC/cryptofuzz-corpora/openssl_latest.zip $OUT/cryptofuzz-openssl_seed_corpus.zip
+fi
+
+##############################################################################
+# Compile Openssl 1.1.0 (without assembly)
+cd $SRC/openssl-OpenSSL_1_1_0-stable/
+make clean || true
+./config --debug no-asm enable-md2 enable-rc5 $CFLAGS
+make depend
+make -j$(nproc)
```
Comment (s-zanella) [inline projects/cryptofuzz/build.sh]:
```suggestion
make -j$(nproc) build_generated libcrypto.a
```

File: projects/cryptofuzz/build.sh [hunk 3]
```diff
@@ -272,3 +272,117 @@ cp $SRC/cryptofuzz/cryptofuzz $OUT/cryptofuzz-boringssl-noasm
...
+cd $SRC;
+unzip OpenSSL_1_0_2-stable.zip
+
+if [[ $CFLAGS != *sanitize=memory* ]]
+then
+    # Compile Openssl 1.0.2 (with assembly)
+    cd $SRC/openssl-OpenSSL_1_0_2-stable/
+    ./config --debug enable-md2 enable-rc5 $CFLAGS
+    make depend
+    make -j$(nproc)
```
Comment (s-zanella) [inline projects/cryptofuzz/build.sh]:
```suggestion
    make -j$(nproc) build_crypto
```

File: projects/cryptofuzz/build.sh [hunk 4]
```diff
@@ -272,3 +272,117 @@ cp $SRC/cryptofuzz/cryptofuzz $OUT/cryptofuzz-boringssl-noasm
...
+    cp $SRC/cryptofuzz-corpora/openssl_latest.zip $OUT/cryptofuzz-openssl_seed_corpus.zip
+fi
+
+##############################################################################
+# Compile Openssl 1.0.2 (without assembly)
+cd $SRC/openssl-OpenSSL_1_0_2-stable/
+make clean || true
+./config --debug no-asm enable-md2 enable-rc5 $CFLAGS
+make depend
+make -j$(nproc)
```
Comment (s-zanella) [inline projects/cryptofuzz/build.sh]:
```suggestion
    make -j$(nproc) build_generated libcrypto.a
```

**Comment by guidovranken** [comment]
Thanks for the fixes @s-zanella . I tried your build speedups but I got errors so I'm leaving them out for now, will try again later.

@jonathanmetzman Travis now fails because "The job exceeded the maximum log length, and has been terminated.". Can't do anything about that now unless you want me to redirect output to /dev/null.

File: projects/cryptofuzz/build.sh [hunk 5]
```diff
@@ -100,15 +100,12 @@ fi
...
 then
     # Compile EverCrypt (with assembly)
-    cd $SRC/
-    tar zxvf hacl-star-evercrypt-v0.1alpha1-bugfix.tar.gz
-    mv hacl-star-evercrypt-v0.1alpha1 evercrypt
-
-    cd $SRC/evercrypt/dist/generic
-    make -j$(nproc) || true
+    cd $SRC/evercrypt/dist
+    make -C portable -j$(npro) libevercrypt.a
```
Comment (s-zanella) [inline projects/cryptofuzz/build.sh]: npro -> nproc

**Comment by s-zanella** [comment]
> Thanks for the fixes @s-zanella . I tried your build speedups but I got errors so I'm leaving them out for now, will try again later.

Strange. What errors do you get?
I built and checked the builds of all fuzzers using the `infra/helper.py` script with these changes and didn't get any errors:
https://github.com/s-zanella/oss-fuzz/commit/e257632102edd7b00eb5142d816673afebf8aa0e

Overall build logs are much shorter. For instance, the log for `--engine libfuzzer --sanitizer memory` goes from 22.9k down to 14.3k lines. You can also silence the output of `unzip` with `unzip -q` to save some more space (~5k lines).

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman Travis now fails because "The job exceeded the maximum log length, and has been terminated.". Can't do anything about that now unless you want me to redirect output to /dev/null.

I'm not sure how to solve this in a general way. Maybe we should add `-Wno-warnings` to `CFLAGS` and `CXXFLAGS` on travis? You can try this yourself if you'd like, or adding `-q` to the unzip and `apt-get` commands (as well as removing the `v` argument to `tar`) should save us some unnecessary output.

If you don't want to go this trouble I can merge and see if it fails to build the old-fashioned way.

**Comment by jonathanmetzman** [comment]
btw, I checked that silencing these particular commands cause appropriate error messages to be printed on error. But if you feel like the verbose output provides value, feel to not silence them and I will merge this patch even though it fails travis (some projects will take too long on travis, so I don't expect every patch to pass on travis).

File: projects/cryptofuzz/build.sh (hunk 1 above)
Comment (guidovranken) [reply projects/cryptofuzz/build.sh]: When I apply this I get errors like

```
include/openssl/bio.h:667:27: warning: declaration of 'struct hostent' will not be visible outside of this function [-Wvisibility]
DEPRECATEDIN_1_1_0(struct hostent *BIO_gethostbyname(const char *name))
                          ^
include/openssl/bio.h:668:1: error: expected function body after function declarator
DEPRECATEDIN_1_1_0(int BIO_get_port(const char *str, unsigned short *port_ptr))
^
include/openssl/bio.h:680:55: warning: declaration of 'union BIO_sock_info_u' will not be visible outside of this function [-Wvisibility]
                  enum BIO_sock_info_type type, union BIO_sock_info_u *info);
                                                      ^
In file included from crypto/aes/aes_wrap.c:10:
In file included from crypto/include/internal/cryptlib.h:26:
include/openssl/bio.h:667:27: warning: declaration of 'struct hostent' will not be visible outside of this function [-Wvisibility]
DEPRECATEDIN_1_1_0(struct hostent *BIO_gethostbyname(const char *name))
                          ^
include/openssl/bio.h:668:1: error: expected function body after function declarator
DEPRECATEDIN_1_1_0(int BIO_get_port(const char *str, unsigned short *port_ptr))
^
include/openssl/bio.h:680:55: warning: declaration of 'union BIO_sock_info_u' will not be visible outside of this function [-Wvisibility]
                  enum BIO_sock_info_type type, union BIO_sock_info_u *info);
                                                      ^
In file included from crypto/aes/aes_ige.c:10:
In file included from crypto/include/internal/cryptlib.h:27:
include/openssl/err.h:249:1: error: expected function body after function declarator
DEPRECATEDIN_1_0_0(void ERR_remove_state(unsigned long pid))
^
In file included from crypto/aes/aes_wrap.c:10:
In file included from crypto/include/internal/cryptlib.h:27:
include/openssl/err.h:249:1: error: expected function body after function declarator
DEPRECATEDIN_1_0_0(void ERR_remove_state(unsigned long pid))
^
crypto/aes/aes_ige.c:47:5: error: use of undeclared identifier 'OPENSSL_FILE'; did you mean 'OPENSSL_die'?
    OPENSSL_assert(in && out && key && ivec);
    ^
include/openssl/crypto.h:332:60: note: expanded from macro 'OPENSSL_assert'
    (void)((e) ? 0 : (OPENSSL_die("assertion failed: " #e, OPENSSL_FILE, OPENSSL_LINE), 1))
```

**Comment by guidovranken** [comment]
@jonathanmetzman all Travis builds pass now.

**Pull request merged**